### ISSUE TITLE
Master hard crash (no backtrace) on Nehalem with ASM enabled

### ISSUE THREAD

**Issue by xiphmont** [xiph/rav1e#1752]
master as of ce13dcdecf5e63b41ec406813ba8c0b938d7cf4e hard crashes here on my development box (nehalem).

 INFO  rav1e::api::config > CPU Feature Level: SSSE3
 INFO  rav1e::api::internal > Using 1 tile
./build.sh: line 51: 29639 Segmentation fault      (core dumped) cargo run --bin rav1e $BUILD_TYPE -- $SEQ -o $ENC_FILE -s 3 -r $REC_FILE

**Comment by shssoichiro** [comment]
This may be the same as https://github.com/xiph/rav1e/issues/1715

**Comment by xiphmont** [comment]
I had pulled since then.  And this is 'build.sh' tanking, not some random file we seldom test.

**Comment by xiphmont** [comment]
let me bisect once I get my AWCY test running....

**Comment by shssoichiro** [comment]
This almost definitely would have started in https://github.com/xiph/rav1e/pull/1713, because before then we never used the SSE2 or SSSE3 optimizations. A Nehalem CPU would have run only native Rust code. And unfortunately that doesn't tell us what change in the ASM actually broke (probably the ASM has been broken forever).

**Comment by ycho** [comment]
> This almost definitely would have started in #1713, because before then we never used the SSE2 or SSSE3 optimizations. A Nehalem CPU would have run only native Rust code. And unfortunately that doesn't tell us what change in the ASM actually broke (probably the ASM has been broken forever).

Then, would #1748 also fix this?

**Comment by shssoichiro** [comment]
The nyan.y4m in the build.sh script is also 480 width, which is not mod 64. So, this corresponds to #1715.

**Comment by shssoichiro** [comment]
@ycho, no, that wouldn't affect this on x86_64 CPUs.

**Comment by xiphmont** [comment]
Yes, I noticed that I was suddenly seeing ASM turned on where it was not before.

**Comment by ycho** [comment]
> @ycho, no, that wouldn't affect this on x86_64 CPUs. Although we could re-disable SSE2 optimizations, but that means non-AVX2 CPUs would go back to being slow.

Ah- #1748 is abut 32 and 64 bit architectures.

**Comment by xiphmont** [comment]
[FTR, Nehalem does have SSSE3, so I agree something that was likely broken forever got turned on]

**Comment by xiphmont** [comment]
Yeah, sadly, bisect told us nothing useful. 45585ee552f3ce43b008006e58979f2ffdea8b8c (#1713) is where things broke, as you guessed.

**Comment by shssoichiro** [comment]
The way I got the backtrace in #1715 was using the rust-lldb tool. It helped determine that the AVX2 implementation of get_sad_16x16 works fine, but the SSE2 implementation is broken on non-mod-64 width videos. There's no SSSE3 implementation for the get_sad functions, so SSSE3 CPUs fall back to the SSE2 implementation.
